flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/StringFunctionsITCase.java

```diff
@@ -44,10 +45,33 @@ Stream<TestSetSpec> getTestSetSpecs() {
                         endsWithTestCases(),
                         printfTestCases(),
                         startsWithTestCases(),
-                        translateTestCases())
+                        translateTestCases(),
+                        concatenateTestCases())
                 .flatMap(s -> s);
     }
 
+    private Stream<TestSetSpec> concatenateTestCases() {
+        return Stream.of(
+                TestSetSpec.forFunction(BuiltInFunctionDefinitions.PLUS)
+                        .onFieldsWithData("Beginning", " Middle ", "Ending")
+                        .andDataTypes(DataTypes.STRING(), DataTypes.STRING(), DataTypes.STRING())
+                        .testResult(
+                                $("f0").concat($("f1")).concat($("f2")),
+                                "concat(concat(f0, f1), f2)",
+                                "Beginning Middle Ending",
+                                DataTypes.STRING())
+                        .testResult(
+                                concat($("f0"), $("f1"), $("f2")),
+                                "concat(f0, f1, f2)",
+                                "Beginning Middle Ending",
+                                DataTypes.STRING())
+                        .testResult(
+                                $("f0").plus($("f1")).plus($("f2")),
+                                "concat(concat(f0, f1), f2)",
```

Comment:
Not related to the PR and just out of curiosity
Would it make sense to have a special rule to translate it into `concat(f0, f1, f2)` if all args are `CHARACTER_STRING`?

Reply:
Generally yes, not sure if we need a rule or should we convert that when translating from table API.